File: leetcode/array/3sum.go
package main

import "fmt"

func threeSum(nums []int) [][]int {
	if len(nums)  < 3 {
		return nil
	}
	sort(&nums, 0, len(nums)-1)
	res := [][]int{}
	for i:=0; i<len(nums)-2 && nums[i] <= 0; i++ {
		target := -nums[i]
		a := i+1
		b := len(nums)-1
		for a < b {
			if nums[a] + nums[b] > target {
				b--
			} else if nums[a] + nums[b] < target {
				a++
			} else {
				res = append(res, []int{nums[i], nums[a], nums[b]})
				for a < b && nums[a+1] == nums[a] {
					a++
				}
				a++
			}
		}
		for i<len(nums)-2 && nums[i] == nums[i+1] {
			i++
		}
	}
	return res
}

func sort(arr *[]int, start int, end int) {
	if start >= end {
		return
	}
	t := (*arr)[start]
	a := start
	b := end
	for a < b {
		for a < b && (*arr)[b] >= t {
			b --
		}
		if a < b {
			(*arr)[a] = (*arr)[b]
		}
		for a < b && (*arr)[a] < t {
			a ++
		}
		if a < b {
			(*arr)[b] = (*arr)[a]
		}
	}
	(*arr)[a] = t
	sort(arr, start, a-1)
	sort(arr, a+1, end)
}

func main() {
	fmt.Print(threeSum([]int{0, 0, 0}))
}

File: python/leetcode/array/3sum.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = list()
        length = len(nums)
        nums.sort()
        i = 0
        while i < length-2 and nums[i] <= 0:
            if i > 0 and nums[i-1] == nums[i]:
                i = i+1
                continue
            j = i+1
            k = length-1
            while j<k and nums[k] >= 0:
                t = nums[i] + nums[j]
                if t > -nums[k]:
                    k = k - 1
                elif t < -nums[k]:
                    j = j + 1
                else:
                    res.append([nums[i], nums[j], nums[k]])
                    j = j+1
                    k = k-1
            i = i+1
        return list(set([tuple(t) for t in res]))

if __name__ == '__main__':
    # i = -100
    # nums = []
    # while i < 100:
    #     nums += [i]
    #     i = i+1
    nums = [-2, 0, 0, 2, 2]
    # print(nums)
    s = Solution()
    print(s.threeSum(nums))
